quinn-udp/src/log.rs

```diff
@@ -0,0 +1,21 @@
+#![allow(unused_imports, unused_macros)]
+
+#[cfg(all(feature = "direct-log", not(feature = "tracing")))]
+pub(crate) use log::{debug, error, info, trace, warn};
```

Comment:
This is still more fragile than necessary. Rather than reexporting macros based on cfg flags, we should define our own macros that have a uniform interface, with cfg-dependent expansions.

Reply:
@Ralith are you suggesting something along the lines of:

``` rust
macro_rules! log_trace {
    ($($arg:tt)*) => {
        #[cfg(all(feature = "direct-log", not(feature = "tracing")))]
        log::trace!($($arg)*);
        #[cfg(feature = "tracing")]
        tracing::trace!($($arg)*);
        #[cfg(not(any(feature = "direct-log", feature = "tracing")))]
        // no-op
        {};
    };
}
```

Both `tracing` as well as `log` are very liberal in what they accept, e.g. both accept any number of _Token Trees_ (i.e. `($arg::tt)*`):

https://docs.rs/log/latest/src/log/macros.rs.html#152-159

https://docs.rs/tracing/latest/src/tracing/macros.rs.html#1823-1830

While we could define our own macros as above, exposing one uniform interface, namely `($arg::tt)*`, that would not actually be any stricter than what this pull request implements today, as `($arg::tt)*` already allows any input already.

Am I missing something? What source of bugs would your suggestion prevent @Ralith?

Reply:
My thinking was that while they both expose very general interfaces, this facade probably shouldn't, because their generalities may not have much in common. Reviewing the docs, though, I guess they actually have a lot of overlap? Maybe it's fine. CI will help in any case, and this is a strict improvement.